badge_number.go

```diff
@@ -0,0 +1,53 @@
+package apns
+
+import "encoding/json"
+
+// BadgeNumber is much a NullInt64 in
+// database/sql except instead of using
+// the nullable value for driver.Value
+// encoding and decoding, this is specifically
+// meant for JSON encoding and decoding
+type BadgeNumber struct {
+	Number uint
+	IsSet  bool
+}
+
+// Unset will reset the BadgeNumber to
+// both 0 and invalid
+func (b *BadgeNumber) Unset() {
+	b.Number = 0
+	b.IsSet = false
+}
+
+// Set will set the BadgeNumber value to
+// the number passed in, assuming it's >= 0.
+// If so, the BadgeNumber will also be marked valid
+func (b *BadgeNumber) Set(number uint) {
+	b.Number = number
+	b.IsSet = true
+}
+
+// MarshalJSON will marshall the numerical value of
+// BadgeNumber
+func (b BadgeNumber) MarshalJSON() ([]byte, error) {
```

Comment:
*As I understand it* although I'm having difficulty finding the appropriate documentation, pointers are dereferenced before they're marshaled.

"Pointers will be encoded as the values they point to (or 'null' if the pointer is nil)."
https://blog.golang.org/json-and-go

So if the marshaller looks for a struct implementation of MarshalJSON and doesn't find it, it defaults to the standard struct encoder.

Here's a demo: https://play.golang.org/p/ZbqaTFXYSp